app/components/channel_link/channel_link.js

```diff
@@ -39,17 +52,55 @@ export default class ChannelLink extends React.PureComponent {
         return null;
     }
 
-    handlePress = () => {
-        this.props.actions.setChannelDisplayName(this.state.channel.display_name);
-        this.props.actions.handleSelectChannel(this.state.channel.id);
+    componentDidMount() {
+        if (!this.state.channel && !this.props.joinableChannel) {
+            const {actions, channelName, currentTeamId} = this.props;
+            actions.searchChannels(currentTeamId, channelName);
+        }
+    }
+
+    handlePress = async () => {
+        const {
+            actions,
+            currentUserId,
+            joinableChannel,
+            onChannelLinkPress,
+        } = this.props;
+
+        let channel;
+        if (joinableChannel) {
+            const result = await actions.joinChannel(currentUserId, joinableChannel.team_id, joinableChannel.id);
 
-        if (this.props.onChannelLinkPress) {
-            this.props.onChannelLinkPress(this.state.channel);
+            if (result.error || !result.data || !result.data.channel) {
+                const joinFailedMessage = {
+                    id: t('mobile.join_channel.error'),
+                    defaultMessage: "We couldn't join the channel {displayName}. Please check your connection and try again.",
+                };
+
+                const utils = require('app/utils/general');
```

Comment:
Why require this dynamically instead of importing it at the top of the file? This component is far enough into the app that I don't think we need to worry about lazily requiring stuff.